src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+                    valueCountLimit,
+                    equalsEncoded,
+                    andEncoded,
+                    out var consumed);
+
+                buffer = buffer.Slice(consumed);
+                return;
+            }
+
+            TryParseValuesSlow(ref buffer,
+                ref accumulator,
+                isFinalBlock,
+                encoding,
+                keyLengthLimit,
+                valueLengthLimit,
+                valueCountLimit,
+                equalsEncoded,
+                andEncoded);
+        }
+
+        // Fast parsing for single span in ReadOnlySequence
+        private static void TryParseFormValuesFast(ReadOnlySpan<byte> span,
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock,
+            Encoding encoding,
+            int keyLengthLimit,
+            int valueLengthLimit,
+            int valueCountLimit,
+            ReadOnlySpan<byte> equalsEncoded,
+            ReadOnlySpan<byte> andEncoded,
+            out int consumed)
+        {
+            ReadOnlySpan<byte> key = default;
+            ReadOnlySpan<byte> value = default;
+            consumed = 0;
+
+            while (span.Length > 0)
+            {
+                var equals = span.IndexOf(equalsEncoded);
+
+                if (equals == -1)
+                {
+                    break;
+                }
+
+                key = span.Slice(0, equals);
+
+                span = span.Slice(key.Length + equalsEncoded.Length);
+
+                var ampersand = span.IndexOf(andEncoded);
+
+                value = span;
+
+                if (ampersand == -1)
+                {
+                    if (!isFinalBlock)
+                    {
+                        // We can't that what is currently read is the end of the form value, that's only the case if this is the final block
+                        // If we're not in the final block, the consume nothing
+                        break;
+                    }
+
+                    // If we are on the final block, the remaining content in value is what we want to add to the KVAccumulator.
+                    // Clear out the remaining span such that the loop will exit.
+                    span = Span<byte>.Empty;
+                }
+                else
+                {
+                    value = span.Slice(0, ampersand);
+
+                    span = span.Slice(ampersand + andEncoded.Length);
+                }
+
+                var decodedKey = GetDecodedString(key, encoding);
+                var decodedValue = GetDecodedString(value, encoding);
+
+                AppendAndVerify(ref accumulator, keyLengthLimit, valueLengthLimit, valueCountLimit, decodedKey, decodedValue);
+
+                // Cover case where we don't have an ampersand at the end.
+                consumed += key.Length + value.Length + (ampersand == -1 ? equalsEncoded.Length : equalsEncoded.Length + andEncoded.Length);
+            }
```

Comment:
There needs to be a check here that if isFinalBlock, make sure we've consumed all of the data. There could be garbage at the end of the form.

Reply:
I don't think I follow. How could there be garbage here? If the encoded bytes had an extra invalid byte at the end?

Reply:
There’s no check required here if you throw in the async loop